File: mobile/lib/features/products/data/product.dart
/// Ürün kartında tanımlanan indirim tipi. `null` = indirim yok.
enum DiscountType {
  percent,
  amount;

  String get apiValue => name;

  static DiscountType? fromApi(String? raw) => switch (raw) {
        'percent' => DiscountType.percent,
        'amount' => DiscountType.amount,
        _ => null,
      };
}

class Product {
  Product({
    required this.id,
    required this.name,
    required this.unit,
    required this.price,
    required this.stock,
    this.description,
    this.imageUrl,
    this.categoryId,
    this.categoryName,
    this.discountType,
    this.discountValue = 0,
    double? effectivePrice,
  }) : effectivePrice = effectivePrice ??
            computeEffectivePrice(price, discountType, discountValue);

  final String id;
  final String name;
  final String? description;
  final String unit;

  /// İndirimsiz liste fiyatı.
  final double price;
  final double stock;
  final String? imageUrl;
  final String? categoryId;
  final String? categoryName;

  final DiscountType? discountType;
  final double discountValue;

  /// Müşterinin ödediği fiyat — indirim düşülmüş hâli. Backend hesaplar,
  /// eski/eksik yanıtlarda yerelde hesaplanır.
  final double effectivePrice;

  bool get hasDiscount => discountType != null && effectivePrice < price;

  /// Satırda gösterilecek indirim rozeti: "%10" ya da "₺5 indirim".
  String? get discountLabel {
    if (!hasDiscount) return null;
    if (discountType == DiscountType.percent) {
      return '%${_trimZeros(discountValue)}';
    }
    return '${_trimZeros(discountValue)} ₺ indirim';
  }

  static String _trimZeros(double v) =>
      v == v.roundToDouble() ? v.toStringAsFixed(0) : v.toString();

  /// Backend'deki `compute_effective_price` ile aynı kural.
  static double computeEffectivePrice(
    double price,
    DiscountType? type,
    double value,
  ) {
    if (type == null || value <= 0) return price;
    final discount = type == DiscountType.percent
        ? price * (value > 100 ? 100 : value) / 100
        : (value > price ? price : value);
    final effective = price - discount;
    return effective < 0 ? 0 : double.parse(effective.toStringAsFixed(2));
  }

  factory Product.fromJson(Map<String, dynamic> json) {
    final cat = json['category'] as Map<String, dynamic>?;
    final price = (json['price'] as num).toDouble();
    final discountType = DiscountType.fromApi(json['discount_type'] as String?);
    final discountValue = (json['discount_value'] as num?)?.toDouble() ?? 0;
    return Product(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      unit: json['unit'] as String,
      price: price,
      stock: (json['stock'] as num).toDouble(),
      imageUrl: json['image_url'] as String?,
      categoryId: cat?['id'] as String?,
      categoryName: cat?['name'] as String?,
      discountType: discountType,
      discountValue: discountValue,
      effectivePrice: (json['effective_price'] as num?)?.toDouble(),
    );
  }
}

File: backend/app/models/product.py
import enum
from decimal import ROUND_HALF_UP, Decimal
from uuid import UUID

from sqlalchemy import Enum, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, UUIDPKMixin

_CENT = Decimal("0.01")


class DiscountType(str, enum.Enum):
    """İndirimin nasıl okunacağı: yüzde mi, sabit TL mi."""

    PERCENT = "percent"
    AMOUNT = "amount"


def compute_effective_price(
    price: Decimal,
    discount_type: "DiscountType | None",
    discount_value: Decimal | None,
) -> Decimal:
    """Ürünün indirim uygulanmış satış fiyatı. Negatife düşmez."""
    if discount_type is None or not discount_value or discount_value <= 0:
        return Decimal(price).quantize(_CENT, rounding=ROUND_HALF_UP)

    price = Decimal(price)
    if discount_type == DiscountType.PERCENT:
        pct = min(Decimal(discount_value), Decimal("100"))
        discount = price * pct / Decimal("100")
    else:
        discount = min(Decimal(discount_value), price)

    effective = price - discount
    if effective < 0:
        effective = Decimal("0")
    return effective.quantize(_CENT, rounding=ROUND_HALF_UP)


class Product(Base, UUIDPKMixin, TimestampMixin):
    """Ürün — toptan perakendenin satışa sunduğu kalem."""

    __tablename__ = "products"

    name: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    unit: Mapped[str] = mapped_column(String(20), nullable=False)  # adet, kg, lt, koli, ...
    price: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    stock: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False, default=Decimal("0"))
    image_key: Mapped[str | None] = mapped_column(String(512), nullable=True)  # S3 key

    # İndirim ürün kartında tanımlanır; her satışta otomatik uygulanır.
    # discount_type NULL = indirim yok (discount_value o zaman 0 tutulur).
    discount_type: Mapped[DiscountType | None] = mapped_column(
        Enum(DiscountType, name="discount_type", values_callable=lambda x: [e.value for e in x]),
        nullable=True,
    )
    discount_value: Mapped[Decimal] = mapped_column(
        Numeric(12, 2), nullable=False, default=Decimal("0"), server_default="0"
    )

    category_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("categories.id", ondelete="SET NULL"), nullable=True
    )
    tenant_id: Mapped[UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False, index=True
    )
    category: Mapped["Category | None"] = relationship(back_populates="products")  # noqa: F821

    @property
    def effective_price(self) -> Decimal:
        """Müşterinin ödeyeceği birim fiyat (indirim düşülmüş)."""
        return compute_effective_price(self.price, self.discount_type, self.discount_value)

    @property
    def has_discount(self) -> bool:
        return self.effective_price < Decimal(self.price)
